File: core/loaders/hakim-loader.mjs
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const VALID_MODES = new Set(['lite', 'full', 'ultra', 'off']);

export function normalizeMode(mode = 'full') {
  const value = String(mode || 'full').trim().toLowerCase();
  return VALID_MODES.has(value) ? value : 'full';
}

export function getCanonicalSkillPath() {
  return path.resolve(__dirname, '..', 'hakim-skill', 'SKILL.md');
}

export function loadSkillContent(skillPath = getCanonicalSkillPath()) {
  return fs.readFileSync(skillPath, 'utf8');
}

export function getModeDirective(mode = 'full') {
  const normalized = normalizeMode(mode);
  const directives = {
    lite: 'Lite mode: execute the request and mention a materially smaller safe alternative when one exists.',
    full: 'Full mode: apply the complete Hakim decision model with proportional verification.',
    ultra: 'Ultra mode: challenge additions, abstractions, and dependencies aggressively while preserving the required outcome and real guards.',
    off: 'Hakim guidance disabled for this session.',
  };
  return directives[normalized];
}

export function getRules(mode = 'full', options = {}) {
  const normalized = normalizeMode(mode);
  if (normalized === 'off') {
    return '# Hakim disabled\n\nHakim guidance is off for this session.\n';
  }
  const skillPath = options.skillPath || getCanonicalSkillPath();
  const skill = loadSkillContent(skillPath);
  return [
    `# Hakim activation (${normalized})`,
    '',
    getModeDirective(normalized),
    '',
    'Canonical Hakim policy loaded from the active distribution.',
    '',
    skill,
  ].join('\n');
}

export default { normalizeMode, getCanonicalSkillPath, loadSkillContent, getModeDirective, getRules };

File: core/loaders/hakim_loader.py
"""Hakim shared Python loader helpers.

Stdlib-only helpers used by tests, packaging commands, and future adapters.
"""
from __future__ import annotations

from pathlib import Path

VALID_MODES = {"lite", "full", "ultra", "off"}


def normalize_mode(mode: str = "full") -> str:
    value = (mode or "full").strip().lower()
    return value if value in VALID_MODES else "full"


def canonical_skill_path() -> Path:
    return Path(__file__).resolve().parents[1] / "hakim-skill" / "SKILL.md"


def load_skill_content(path: Path | None = None) -> str:
    return (path or canonical_skill_path()).read_text(encoding="utf-8")


def get_mode_directive(mode: str = "full") -> str:
    directives = {
        "lite": "Build what is asked, then name the lazier alternative in one line.",
        "full": "Enforce the Hakim ladder with reuse, stdlib, native platform features, and shortest safe diffs.",
        "ultra": "YAGNI extremist mode: delete before adding and ship the minimum safe change.",
        "off": "Hakim guidance disabled for this session.",
    }
    return directives[normalize_mode(mode)]


def get_rules(mode: str = "full", path: Path | None = None) -> str:
    normalized = normalize_mode(mode)
    if normalized == "off":
        return "# Hakim disabled\n\nHakim guidance is off for this session.\n"
    return "\n".join([
        f"# Hakim activation ({normalized})",
        "",
        get_mode_directive(normalized),
        "",
        "Canonical source: core/hakim-skill/SKILL.md",
        "",
        load_skill_content(path),
    ])
